README.md
# CacheMe

CacheMe is a Geocaching application. See http://www.harbaum.org/till/cacheme



debian/changelog
cacheme (0.9.4-1) unstable; urgency=low

  * 

 -- Till Harbaum <till@harbaum.org>  Mon, 13 Aug 2012 18:56:24 +0200

cacheme (0.9.3-1) unstable; urgency=low

  * Fix multi-line logs
  * Fix log date
  * Allow to click on waypoints
  * search for waypoint names
  * search for gpx memory leak (none found)
  * "publish" option for logs
  * czech translation

 -- Till Harbaum <till@harbaum.org>  Thu, 31 May 2012 21:46:55 +0200

cacheme (0.9.2-1) unstable; urgency=low

  * Live log
  * (spoiler) images
  * exclude own/found from search results
  * option to reload details
  * option to center map on waypoint
  * OpenCaching plugin
  * Don't hardcode geocaching.com for browser logging
  * fix kinetic scrolling

 -- Till Harbaum <till@harbaum.org>  Tue, 15 May 2012 19:06:21 +0200

cacheme (0.9.1-1) unstable; urgency=low

  * Fix waypoint display
  * Fix waypoint description
  * Button for re-authentication
  * info box before browser is being launched
  * simplify plugin interfaces
  * generic info interface
  * generic io requests
  * generic io replies
  * support search
  * map search results
  * increase info cache to 25 entries
  * fix attribute handling
  * kinetic scrolling

 -- Till Harbaum <till@harbaum.org>  Thu, 12 Apr 2012 18:49:53 +0200

cacheme (0.9.0-1) unstable; urgency=low

  * Remove GcBrowser plugin
  * Add help page

 -- Till Harbaum <till@harbaum.org>  Thu, 12 Apr 2012 07:04:43 +0200

cacheme (0.8.17-1) unstable; urgency=low

  * Implement GcLive plugin

 -- Till Harbaum <till@harbaum.org>  Thu, 08 Mar 2012 20:19:52 +0100

cacheme (0.8.16-1) unstable; urgency=low

  * Move motd into tools menu
  * Yet another adaption to a gc.com web site change

 -- Till Harbaum <till@harbaum.org>  Thu, 23 Feb 2012 19:25:46 +0100

cacheme (0.8.15-1) unstable; urgency=low

  * "message of the day" tool
  * Adopt to new map from 02/14/12:
  * request all overview tiles
  * adopt to zoom level
  * request map.tile as this also generates the .info part
  * average geolocation from map grid
  * parse new info stucture
  * remove <strike> tags
  * restrict zoom levels to 13

 -- Till Harbaum <till@harbaum.org>  Sun, 19 Feb 2012 21:18:57 +0100

cacheme (0.8.14-1) unstable; urgency=low

  * Version increase since OVI store doesn't seem to cope with 0.8.13.1

 -- Till Harbaum <till@harbaum.org>  Mon, 13 Feb 2012 14:30:08 +0100

cacheme (0.8.13.1-1) unstable; urgency=low

  * Re-enable icon click workaround for symbian

 -- Till Harbaum <till@harbaum.org>  Sat, 04 Feb 2012 20:11:59 +0100

cacheme (0.8.13-1) unstable; urgency=low

  * tool infrastructure
  * rot/rot13 tool
  * GPS and map center selectable in coordinate tool
  * Adjust/reduce settings dialog spacing
  * coordinate projection tool
  * fix "many caches" cpu load problem on mobile

 -- Till Harbaum <till@harbaum.org>  Tue, 22 Nov 2011 18:45:29 +0100

cacheme (0.8.12-1) unstable; urgency=low

  * replace settings tabs with scrollarea on maemo5
  * option to disable map screen saver
  * avoid qgroupboxes on symbian due to problems with qt 4.7.4 there
  * suppress further background processing for qt 4.7.4
  * show gps precision in nav view
  * support imperial units for distances
  * fixed bug in flick scroller
  * Symbian icon size adjustment
  * Fix user name truncation in Gcbrowser log parsing

 -- Till Harbaum <till@harbaum.org>  Sat, 05 Nov 2011 21:48:12 +0100

cacheme (0.8.11-1) unstable; urgency=low

  * Prepare for official symbian^1 release
  * change settings dialog to cope with bigger fonts (e.g. 5250)
  * obfuscate password in conf file
  * Cache map update requests in background to cope with belle crashes
  * help link in about box
  * Support for QtMobility 1.2.1 map types

 -- Till Harbaum <till@harbaum.org>  Mon, 24 Oct 2011 21:03:32 +0200

cacheme (0.8.10-1) unstable; urgency=low

  * Updated fremantle desktop icons
  * Fixed Waypoint name length problem in nav view
  * Option to disable found/owned caches
  * Tabbed settings dialog
  * Completely disable use of network disk cache

 -- Till Harbaum <till@harbaum.org>  Sun, 09 Oct 2011 20:22:51 +0200

cacheme (0.8.9-1) unstable; urgency=low

  * Adopted gcbrowser login to cope with latest gc.com site changes
  * Catch "UnknownNetworkError" under Symbian
  * Anna icon

 -- Till Harbaum <till@harbaum.org>  Fri, 07 Oct 2011 21:31:51 +0200

cacheme (0.8.8-1) unstable; urgency=low

  * Waypoints can be selected on the map
  * Option to copy coordinates to clipboard
  * Disable screen saver in navigation window
  * Fixed super-stupid bug in log month parsing

 -- Till Harbaum <till@harbaum.org>  Wed, 31 Aug 2011 19:09:35 +0200

cacheme (0.8.7-1) unstable; urgency=low

  * Bugfix in waypoint selection in nav view
  * Fixed premature start in gpsdgeoinfosource
  * Update type icon in compass view

 -- Till Harbaum <till@harbaum.org>  Wed, 24 Aug 2011 18:17:03 +0200

cacheme (0.8.6-1) unstable; urgency=low

  * Local waypoint handling
  * Changed Cloudmade cache handling
  * re-implement pixmap handling to cope with offsets

 -- Till Harbaum <till@harbaum.org>  Wed, 10 Aug 2011 17:42:11 +0200

cacheme (0.8.5-1) unstable; urgency=low

  * Local Waypoint Database
  * Coordinate tool
  * Google maps integration
  * Finger scrolling under Symbian
  * Fixed hint parsing for GPX files

 -- Till Harbaum <till@harbaum.org>  Mon, 1 Aug 2011 19:22:43 +0200

cacheme (0.8.4-1) unstable; urgency=low

  * Re-enabled scrollbar in settings dialog for Maemo5 (only)

 -- Till Harbaum <till@harbaum.org>  Sat, 30 Jul 2011 21:28:51 +0200

cacheme (0.8.3-1) unstable; urgency=low

  * Shrinked config dialog
  * Added option to increase icon size (for E6)
  * Fixed "owned waypoints" crash

 -- Till Harbaum <till@harbaum.org>  Sun, 28 Jul 2011 20:32:11 +0200

cacheme (0.8.2-1) unstable; urgency=low

  * Removed unused QAction in main menu

 -- Till Harbaum <till@harbaum.org>  Sun, 17 Jul 2011 23:38:40 +0200

cacheme (0.8.1-1) unstable; urgency=low

  * Bug fixes in coordinate tool


src/cache.h
#ifndef CACHE_H
#define CACHE_H

#if !defined(Q_OS_SYMBIAN) && !defined(Q_WS_MAEMO_5)
#if QT_VERSION < 0x040800  // doesn't seem to work with qt 4.8
#define STAGING   // define this to test via staging
#endif
#endif

#include <QLandmark>
#include <QDate>

class QXmlStreamWriter;
class QXmlStreamReader;
class QXmlStreamAttributes;

QTM_USE_NAMESPACE

class Description {
 public:
  Description();
  friend QDebug operator<<(QDebug, const Description &);
  void set(bool, const QString &);
  QString text() const;
  bool isHtml() const;
  bool isSet() const;
  void save(QXmlStreamWriter *, const QString &);
  void load(QXmlStreamReader &);

 private:
  QString m_text;
  bool m_isHtml;
};

class Image {
 public:
  Image();
  friend QDebug operator<<(QDebug, const Image &);
  void setName(const QString &);
  void setDescription(const QString &);
  void setUrl(const QString &);
  QString name() const;
  QString description() const;
  QString url() const;
  void save(QXmlStreamWriter *);
  bool load(QXmlStreamReader &);

 private:
  QString m_name, m_description;
  QString m_url;
};

class Waypoint : public QLandmark {
 public:
  enum Type {
    Unknown, 
    Multistage, Parking, Final, Question, Trailhead, Refpoint,
    Custom
  };
  
  Waypoint();
  friend QDebug operator<<(QDebug, const Waypoint &);
  void setComment(const QString &);
  QString comment() const;
  Type type() const;
  void setType(const Type &);
  QString iconFile() const;
  void save(QXmlStreamWriter *);
  bool load(QXmlStreamReader &);

 private:  
  Type m_type;
  QString m_comment;
};

class Attribute {
 public:
  enum Id {
    Unknown = 0, 
    Dogs = 1, Fee = 2, Rappelling = 3, Boat = 4, Scuba = 5, Kids = 6, 
    OneHour = 7, Scenic = 8, Hiking = 9, Climbing = 10, Wading = 11, 
    Swimming = 12, Available = 13, Night = 14, Winter = 15, Cactus = 16, 
    PoisonOak = 17, Snakes = 18, Ticks = 19, Mine = 20, Cliff = 21, 
    Hunting = 22, Danger = 23, WheelChair = 24, Parking = 25, Public = 26, 
    Water = 27, Restrooms = 28, Phone = 29, Picnic = 30, Camping = 31, 
    Bicycles = 32, Motorcycles = 33, Quads = 34, Jeeps = 35, Snowmobiles = 36, 
    Horses = 37, Campfires = 38, Thorn = 39, Stealth = 40, Stroller = 41, 
    FirstAid = 42, Cow = 43, Flashlight = 44, LandF = 45, RV = 46, 
    FieldPuzzle = 47, UV = 48, Snowshoes = 49, Skiis = 50, STool = 51, 
    NightCache = 52, ParkNGrab = 53, AbandonedBuilding = 54, HikeShort = 55, 
    HikeMed = 56, HikeLong = 57, Fuel = 58, Food = 59, WirelessBeacon = 60,
    PartnerShip = 61, Seasonal = 62, TouristOk = 63, TreeClimbing = 64,
    FrontYard = 65, TeamWork = 66
  };
  
  Attribute(const Id &, bool);
  friend QDebug operator<<(QDebug, const Attribute &);
  QString toString() const;
  QStringList iconFile() const;
  bool isOn() const;
  Id id() const;

 private:
  Id m_id;
  bool m_yes;
};

class Log {
 public:
  class Type {
  public:
    enum Id {
      Unknown, 
      Found, NotFound, Maintenance, WriteNote, ReviewerNote,
      EnableListing, PublishListing, WillAttend, Attended,
      Photo, TempDisable, NeedsMaintenance, UpdatedCoordinates,
      Unarchive, NeedsArchived, Archive
    };

    Type();
    void set(Type::Id);
    bool is(Type::Id) const;
    Type::Id id() const;
    static QString iconFile(Type::Id);
    QString iconFile() const;
    static QString toString(Type::Id);
    QString toString() const;

  private:
    Id m_id;
  };

  Log();
  void setType(const Type &);
  void setType(Type::Id);
  void setDate(const QDate &);
  Type type() const;
  QDate date() const;
  void setFinder(const QString &);
  QString finder() const;
  void setDescription(const Description &);
  Description description() const;
  void save(QXmlStreamWriter *);
  bool load(QXmlStreamReader &);
  friend QDebug operator<<(QDebug, const Log &);

 private:
  Type m_type;
  QDate m_date;
  QString m_finder;
  Description m_description;
};

class Tribool {
 public:
  Tribool();
  friend QDebug operator<<(QDebug, const Tribool &);
  void set(bool);
  bool value() const;
  bool isSet() const;
  bool isSetAndTrue() const;
  void saveAttribute(QXmlStreamWriter *, const QString &);
  void loadAttribute(QXmlStreamAttributes &, const QString &);

 private:
  bool m_value;
  bool m_set;
};

class Rating {
 public:
  Rating();
  friend QDebug operator<<(QDebug, const Rating &);
  void set(qreal);
  qreal value() const;
  QString toString() const;
  bool isSet() const;
  void save(QXmlStreamWriter *, const QString &);
  void load(QXmlStreamReader &);
  
 private:
  int m_value;
};

class Container {
 public:
  enum Type {
    ContainerUnknown, ContainerRegular, ContainerSmall, ContainerMicro,
    ContainerOther,   ContainerNotChosen, ContainerLarge, ContainerVirtual
  };
  
  Container();
  friend QDebug operator<<(QDebug, const Container &);
  void set(Type);
  Type get() const;
  static QString toString(Type);
  QString toString() const;
  bool isSet() const;
  static QString iconFile(Type, bool = false);
  QString iconFile(bool = false) const;
  void save(QXmlStreamWriter *, const QString &);
  void load(QXmlStreamReader &);
 
 private:
  Type m_value;
};

class Cache : public QLandmark {
 public:

  enum Type {
    TypeUnknown, TypeTraditional, TypeMulti, TypeMystery, TypeVirtual,
    TypeWebcam, TypeEvent, TypeLetterbox, TypeEarthcache, TypeWherigo,
    TypeMegaEvent, TypeCito };
  
  Cache();
  ~Cache();
  friend QDebug operator<<(QDebug, const Cache &);

  // functions dealing with generic flags
  void setOwned(bool);
  Tribool owned() const;

  void setFound(bool);
  Tribool found() const;

  void setAvailable(bool);
  Tribool available() const;

  // function dealing with owner
  void setOwner(const QString &);
  QString owner() const;

  // functions dealing with the cache type
  void setType(Type);
  Type type() const;
  QString typeString() const;
  static QString typeString(Type);
  QString typeIconFile() const;
  static QString typeIconFile(Type);
  QString overlayIconFile() const;

  // functions dealing with the geocaching.com id
  void setId(int);
  int id() const;

  // functions dealing with the caches uuid
  void setGuid(const QString &);
  QString guid() const;

  // functions dealing with difficulty and terrain ratings
  void setDifficulty(qreal);
  Rating difficulty() const;

  void setTerrain(qreal);
  Rating terrain() const;

  // functions dealing with the container
  void setContainer(Container::Type);
  Container container() const;

  // functions dealing with date of placement
  void setDateOfPlacement(const QDate &);
  QDate dateOfPlacement() const;
  QString dateOfPlacementString() const;

  // functions dealing with descriptions
  void setShortDescription(const Description &);
  Description shortDescription() const;

  void setLongDescription(const Description &);
  Description longDescription() const;

  // functions dealing with the hint
  void setHint(const Description &);
  Description hint() const;
 
  // functions dealing with logs
  QList<Log> logs() const;
  void appendLog(const Log &);
  void clearLogs();

  // functions dealing with attributes
  QList<Attribute> attributes() const;
  void appendAttribute(const Attribute &);
  void clearAttributes();

  // functions dealing with waypoints
  QList<Waypoint> waypoints() const;
  void appendWaypoint(const Waypoint &);
  void updateWaypoint(const Waypoint &);
  void clearWaypoints();

  // functions dealing with images
  QList<Image> images() const;
  void appendImage(const Image &);
  void clearImages();

  // functions for writing and restoring
  bool save(QXmlStreamWriter *);
  bool load(QXmlStreamReader &);
  bool loadWaypoints(QXmlStreamReader &);
  bool loadLogs(QXmlStreamReader &);
  bool loadImages(QXmlStreamReader &);

 private:
  Tribool m_owned, m_found, m_available;
  Type m_type;
  Rating m_terrain, m_difficulty;
  int m_id;
  Container m_container;
  QString m_owner;
  QDate m_dateOfPlacement;
  QString m_guid;
  Description m_shortDescription, m_longDescription;
  Description m_hint;
  QList<Log> m_logs;
  QList<Attribute> m_attributes;
  QList<Waypoint> m_waypoints;
  QList<Image> m_images;
};

#endif // CACHE_H
